demos/embeddings/deep-graph-infomax-cora.ipynb

```diff
@@ -6,7 +6,7 @@
    "source": [
```

Comment:
<p>Maybe GAT is struggling with a small number of attention heads (<code>attn_heads</code> defaults to 1, I think)?</p><p> </p><p>My original suggestion was to write each section out as its own cells to make these sort of per-model tweaks easier. One approach would be to have a function that takes a model and a generator, e.g.</p><p> </p><blockquote>The process for training with Deep Graph Infomax is almost identical for every model, requiring only the model and generator to be wrapped in <code>DeepGraphInfomax</code>  and <code>CorruptedGenerator</code>  respectively. We can define a function to do this for us, to make it easy to train different models for a comparison. The StellarGraph library ensures that the process for predicting with each different model is similar (using the <code>flow</code>  methods), so the functions can include the evaluation code too. The only per-model customisation required in this function is removing the batch dimension for full batch methods like GCN, GAT and APPNP.</blockquote><pre class="ql-syntax" spellcheck="false">def run_deep_graph_infomax(base_model, generator):
&nbsp; &nbsp; corrupted_generator = CorruptedGenerator(generator)
&nbsp; &nbsp; gen = corrupted_generator.flow(G.nodes())

&nbsp; &nbsp; infomax = DeepGraphInfomax(base_model)
&nbsp; &nbsp; x_in, x_out = infomax.in_out_tensors()

&nbsp; &nbsp; model = Model(inputs=x_in, outputs=x_out)
&nbsp; &nbsp; model.compile(loss=tf.nn.sigmoid_cross_entropy_with_logits, optimizer=Adam(lr=1e-3))

&nbsp; &nbsp; history = model.fit(gen, epochs=epochs, verbose=0, callbacks=[es])

&nbsp; &nbsp; x_emb_in, x_emb_out = base_model.in_out_tensors()

&nbsp; &nbsp; # for full batch models, squeeze out the batch dim (which is 1)
&nbsp; &nbsp; if isinstance(base_model, (GAT, GCN, APPNP)):
&nbsp; &nbsp; &nbsp; &nbsp; x_emb_out = tf.squeeze(x_emb_out, axis=0)

&nbsp; &nbsp; emb_model = Model(inputs=x_emb_in, outputs=x_emb_out)

&nbsp; &nbsp; test_gen = generator.flow(test_subjects.index)
&nbsp; &nbsp; train_gen = generator.flow(train_subjects.index)

&nbsp; &nbsp; test_embeddings = emb_model.predict(test_gen)
&nbsp; &nbsp; train_embeddings = emb_model.predict(train_gen)

&nbsp; &nbsp; lr = LogisticRegression(multi_class="auto", solver="lbfgs")
&nbsp; &nbsp; lr.fit(train_embeddings, train_subjects)

&nbsp; &nbsp; y_pred = lr.predict(test_embeddings)
&nbsp; &nbsp; acc = (y_pred == test_subjects).mean()

&nbsp; &nbsp; return acc
</pre><p> </p><p>and then it can be used for each of them like follows (with this we wouldn't have to re-run GCN and could reuse the full-batch generator):</p><p> </p><pre class="ql-syntax" spellcheck="false">gat_model = GAT(layer_sizes=[512], activations=["relu"], attn_heads=8, generator=generator)
gat_acc = run_deep_graph_infomax(gat_model, generator)
</pre><p> </p><pre class="ql-syntax" spellcheck="false">appnp_model = APPNP(layer_sizes=[512], activations=["relu"], generator=generator)
appnp_acc = run_deep_graph_infomax(appnp_model, generator)
</pre><p> </p><pre class="ql-syntax" spellcheck="false">graphsage_generator = GraphSAGENodeGenerator(G, batch_size=2700, num_samples=[5])
graphsage_model = GraphSAGE(layer_sizes=[512], activations=["relu"], generator=generator)
graphsage_acc = run_deep_graph_infomax(graphsage_model, generator)
</pre><p> </p><p>And then the final results would use the various <code>*_acc</code> variables.</p><p> </p><p>(It might make sense to rename the GCN <code>generator</code> to <code>full_batch_generator</code>, if we go with this.)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1162/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Reply:
I reafctored the notebook like you suggested, looks much nicer 👍 I decreased layer size and epochs here to make the whole thing run abit faster and because I didn't notice much of a drop in performance. 

I'm still getting the low accuracy for GAT w/ the larger number of heads tho :(

Reply:
aside from this results issue I think its ready for a re-review